{
 "metadata": {
  "name": "",
  "signature": "sha256:77443d952ce3eaf5f55b51b74161615c85078d30d898c13c181556124c63a30a"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# import libraries we're using\n",
      "from Bio.Seq import Seq\n",
      "from Bio.SeqRecord import SeqRecord\n",
      "from Bio import SeqIO\n",
      "from intermine.webservice import Service\n",
      "\n",
      "\n",
      "# define global variables\n",
      "HomologyLength=1000 \n",
      "\n",
      "def make():\n",
      "    service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
      "    template = service.get_template('Gene_GenomicDNA')\n",
      "\n",
      "    rows = template.rows(\n",
      "        E = {\"op\": \"LOOKUP\", \"value\": GeneName, \"extra_value\": \"S. cerevisiae\"}\n",
      "    )\n",
      "    \n",
      "    # this service seems to return multiple similar genes but we want the first one only, so count\n",
      "    # and it returns information about the gene you want\n",
      "    count=0\n",
      "    for row in rows:\n",
      "        \n",
      "        count=count+1\n",
      "        if count==1:\n",
      "            descr= row[\"description\"]\n",
      "            GeneSeq=Seq(row[\"sequence.residues\"])\n",
      "            GeneSysName=row[\"secondaryIdentifier\"]\n",
      "            print(\" \")\n",
      "            print(\"I think you want...... \"+row[\"secondaryIdentifier\"])\n",
      "            print(row[\"description\"])\n",
      "            print(\" \")\n",
      "            print(row[\"sequence.residues\"])\n",
      "            print(\" \")\n",
      "            print(\"Good choice! I have a feeling you're going to get lucky with this one.\")\n",
      "            print(\"Give me a second to put some of my ducks in a circle.\")\n",
      "            print(\" \")\n",
      "\n",
      "            \n",
      "    #let's create a record for the oldGene\n",
      "    oldGene = SeqRecord(GeneSeq, id=GeneSysName)\n",
      "\n",
      "def fetchNeighbor(myGeneRecord,type,distance):\n",
      "    print(\"Wait a minute I'm not that smart yet.\")\n",
      "    \n",
      "    \n",
      "    # the systematic name (saved as feature) contains information about the location of the gene\n",
      "    # index 1 points to a letter corresponding to the chromosome number\n",
      "\n",
      "    if myGeneRecord.features[1]==\"A\":\n",
      "        myChromosome=SeqIO.read(\"Scer01.fasta\", \"fasta\")\n",
      "        # Chromosome is a SeqRecord\n",
      "    \n",
      "    #Let's explicitely name the sequences\n",
      "    myGeneSeq=myGeneRecord.seq\n",
      "    myChromosomeSeq=myChromosome.seq\n",
      "    \n",
      "    #\n",
      "    if myChromosomeSeq.find(myGeneSeq)==-1:\n",
      "        print (\"I flipped your chromosome so you don't get confused.\")\n",
      "    \n",
      "    StartIndex = myChromosomeSeq.find(myGeneSeq)\n",
      "    EndIndex = len(myGeneSeq)\n",
      "    \n",
      "    print(StartIndex)\n",
      "    print(EndIndex)\n",
      "    \n",
      "    print(myChromosomeSeq[StartIndex:EndIndex+1])\n",
      "\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}