{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hi baby, what are you in the mood for today? Using quotes enter\n",
      "\"1\" to insert to an empty locus\n",
      "\"2\" to edit an existing gene\n",
      " \n",
      "Your answer: 2\n",
      " \n",
      "OK sweetie. Which locus do you want to edit, e.g., \"OAF1\": \n",
      "Right now I only have chromosome 1 for testing.\n",
      "Your answer: \"OAF1\"\n",
      " \n",
      "I think you want...... YAL051W\n",
      "Oleate-activated transcription factor; subunit of a heterodimeric complex with Pip2p, which binds to oleate-response elements (ORE) in the promoter of genes involved in beta-oxidation of fatty acids, peroxisome organization and biogenesis, activating transcription in the presence of oleate; regulates chromatin silencing at telomeres; involved in diauxic shift; OAF1 has a paralog, PIP2, that arose from the whole genome duplication\n",
      " \n",
      "ATGGTCGAAAATAGTACGCAGAAGGCCCCACATGCCGGAAATGATGATAATAGCTCTACCAAGCCATATTCGGAGGCGTTTTTCTTAGGGTTCAATAACCCAACGCCTGGATTAGAAGCTGAGCACTCAAGCACATCGCCTGCCCCCGAGAACTCCGAAACACATAATAGGAAAAGAAATAGAATATTGTTTGTCTGCCAGGCTTGTAGGAAGTCAAAAACAAAGTGTGATAGAGAAAAACCTGAATGTGGTCGATGCGTCAAGCATGGGTTAAAATGTGTTTATGACGTATCAAAACAGCCAGCACCACGAATTCCGAGTAAAGACGCCATTATATCAAGGTTGGAAAAAGATATGTTTTATTGGAAAGATAAAGCTATGAAGCTACTAACAGAGAGAGAGGTGAATGAATCAGGCAAGAGATCAGCAAGTCCGATCAATACAAACAATGCTAGCGGGGACAGTCCTGATACCAAGAAGCAGCATAAAATGGAACCTATATATGAACAAAGTGGTAACGGGGATATAAACAATGGTACCAGAAATGATATTGAAATCAACTTGTATAGAAGTCATCCAACCATGATCATGAGTAAAGTCATGAAAAGAGAAGTTAAGCCGTTATCTGAAAATTATATTATAATTCAGGACTGTTTTCTAAAAATCCTGGTCACTTCAGTGTTCCTTGACACTTCAAAGAACACGATGATACCGGCATTGACGGCAAACGCGAATATTACAAGAGCCCAGCCTAGCGTAGCAAATAACCTTTTGAAATTGAAGGAAATGCTAATCAGACAGTGTCAAACCGAAGATGAAAAAAATCGTGTAAACGAATTCACTGATAGAATACTACAAAATACAAATTCAAATAGAAACTTGAAAATCGGTATGCTATTATCAATGCTTTACAATTCTGTCGGTTACCAATATCTGGAGGATCATTGCCCTCAAGGTGGCGAATATTCGGATTTATTGAGAAATTTGATCAATGAATGTGAAGCTATTTTGCCATCTTACGAAATTATTGAACGCTACAAGAACCACTTTTATGAGTACGTTTATCCAAGTCTACCTTTCATCGAATTAGAAATTTTTGAAGAATCATTAAGTCAAACAATTTTTCCGGACCCAAACAACCCCTCCAAGGTGCAAATACGTATGGGTAGCACACATTTGAGAGCTAAGGTGGAAAACTTGAGTCTTCTATTGGTTATCTTGAAACTCTCATACATGTCAATAAGGTTTTTAGATCATAGTACAGCAGACTCGAGTTTTTATCTTTCAAAGGAAATAATTGATAAATACCCAATACCGAACGATTTTATTTTATTGAGTCAAAGATGTCTAGCATCGGAAAATTGGTGTGCATGCGCTAATGAAAACATCATATCATGTTTACTATATATCTGGTCTTTTTTTGCTTTTTCTCCTGAAGAGGGTGATTTCTTTCTCGAGCACCCCACCGATGTTATCAGTAGTTTGATAATGATGCTTTCCACCTCGATTGGTCTCCACAGAGATCCTTCAGATTTCCCTCAATTGATTTCCCCGTCCACCTCAGATAAAAGAACCTTGAATCACAGAAGAATACTCTGGTTGAGTATCGTTACCGTTTGTTCGTTTGAAGCAAGTCTCAAAGGTAGACATTCTGTCTCACCGATATCTTTAATGGCCTTATTCCTAAATATTAAGGATCCTGATTCTCTGACGGTATATATGAACCGAGTTAGGGGCGATCTAAGCGATATCAATAATCACAAACTTTTGAGAATTCATAAATTTACATTCAAGAGAGCCCAGCTTGCGTTACTCCTGTCGGACTTAGATAACTTGACGATGACATACTATGGTAGTTTCCATTTGCATTCAATTGAATTCATAAGAGAAAAGATTGAGATTTTTGTGGAGGAAAACTTTCCCATAGTACCATTAAAAAGTGTCGCACAGGATAAGTCAGACCTTGATGACATGAATGTGATTTCAGAAATGAATATATTATCTTCAGAAAATTCTTCTTCATTTCACAATCGAATAATGAATAAACTATTGATGTTGAGAACTTCAATGGCCGTATTCTTGCATTTTGAAACACTTATCACTAAGGATAAAAGTATCTTCCCATTCTACAAGAAATACTTTATGGTTAGCTGTATGGATGCGTTGTCACTAATAAATTATTTCAATAAGTTTTTCAACGGAGAATATCGACACGCAATATCTTCTTTAACCAGTTTTAATGTTACAAAATTTATTCAGTTAGCACTATCCAGCACAATCTTCAGCCTATTAGGGATTATACTAAGAATAGGTTTAGCCATCCATATGTTATCTTCTGAAGTACAAAAGTTATCGGGAACGACAGATCCAAGAATAAAGGAGTTAAATACCAAAGTCGAAAAATTTAGTACCCTGCAAAGAGATCTCGAGTCTGCTTTAGAAGGTATATATTGCTCTGCTTCGGAACATTTAAGATTCACATACTTCCCCGTTTTTAAGATGTTGGCTTTATTCGATGTCATTGTACAAAGGATGAGAAAGGGTGAATTATGGCACGGCATATTTACGATGATTCAAATGGAACAAATGCATTCTAGGATAATCAAGACATTAAGCATTACCTTAGGAGTCAAACTGGACAAAAAGGATAGGCTATTAGAGGAATTGATGGCATGCAATCACGTTGCGAATTTTAGCGTTGAAGATATAGATGAGCTGAACCGTAATATCAAAAAAGAGATTCAAATTTCTTCAGGATTGAAGCCGCCTGTAAACACAATTGACTTAACCAACGGCGAACCATTCGGAAATGCTGTTCCTACCTTCACAAAGACATGGAGTTCATCCTTAGATAATTTAGAAAAACTATCATCGGCCGCTGCAGTTGGTCAGAGCTTGGACTACAACAGTGGTTTACGTCAGGGTCCTTTGGCGGGTGGTGGTTCAAAAGAGCAAACGCCAATAGCCGGGATGAATAACTTGAACAATTCAATCAATGCTACACCAATTGTCGATAACTCATCTGGATCACAACTTCCTAATGGTTTCGATAGAGGCCAAGCGAATAATACTCCTTTTCCAGGTTATTTTGGAGGTTTGGATTTATTTGATTATGACTTTTTGTTTGGCAATGACTTTGCTTAA\n",
      " \n",
      "Good choice! I have a feeling you're going to get lucky with this one.\n",
      " \n",
      "Give me a second to put some of my ducks in a circle...\n",
      " \n",
      "I'm fetching upstream and downstream homology reghhions for you...\n",
      " \n",
      "OK I found them.\n",
      " \n",
      "UpstreamHomology reghhion is\n",
      "GTGAATTTGATCAATGCATTTTTCTTTTTGTTTTTCAGTAATTTGTTCAAGCAACCAGCAGTGGTTTCGTCAGTGATGGCGGTTATTCTGTTCGTTTTGAACGCGGTGTTTGCTCTATTCCTATTATTGTTCACTATTGTCACCTGTACACTGGCATTACTACACAGAAACCCAGATGTCCGTTACCAACCAATGAAAGATGACCGTGTGTCATTCATTCCTAAGATTCAAAATGATTTCGATGGCAAAAACAAAAATGATTCTGAACTGTTTGAATTGAGAAAAGCTGTTATGGACACCAATGAAAATGAGGAAGAAAAAATGTTCCGTGACGACACTTTCGGCAAGAACCTGAATGCAAACACAAATACAGCAAGACTCTTTGATGATGAGACTAGTTCATCCTCTTTTAAGCAAAATTCCTCTCCCTTCGATGCCTCGGAAGTAACGGAGCAACCTGTGCAACCAACCTCCGCTGTCATGGGTACGGGTGGCAGCTTCTTGTCTCCACAGTACCAACGTGCGTCATCTGCTTCTCGTACTAATCTAGCGCCGAATAATACAAGCACCTCCAGTTTAATGAAGCCTGAATCAAGTCTCTACCTGGGGAATTCCAATAAATCATATTCGCATTTTAACAACAACGGCAGCAACGAAAACGCCCGCAACAACAACCCATATTTGTAATCCAATATATACTCACATGTAACAACTTATTATATAAATATTTAAGGGCAAGGATATCCTACATTATATTTCATAGAAAACCGCTCAAAAAGGTGTATTATCTCCATTACATCCCAACACCACACATATTTCAGCGATAAAAACCTTAAATGTGAAATTCGCTTTGGCTCTGCTTCCTTAAATGTACGCAATTGCCGCTTTTTTCTGACATCTTTTTTGACGTGTAGAGAAGGAAACAGATCCTCCAGAAGGGATTTACTGTTGGCTATTTTGTGTTAGAAGCAGGTTAATAATAGATTAGGTTGCGTAAGTC\n",
      " \n",
      "DownstreamHomology reghhion is\n",
      "AAATTTTCTTTCCAAACTCCTACCTATTCATTTCATCAATTAATTAATATTATATAGCCACGAATTTATGAAACTGACCGATAATATAAAGTGCTCAATATATATATATATGTATATAACGGTTAACGTAAGAAGAGCTCTTCCCTCTTAAACATTCGAAAAATGATTGAACCAGTATATTTGGTCGAGCAAGACTTTCTCCTTCGCATATTTTACGGCAGGTATGGATATATCGCCTCTTGCTGCAAACCCGTGAGCCACACCACTGAAGAGGTCTAACTGGTAAGTAGCGTGATTATCCTTTAATTTTTCCTCCGTTAAGTGTCTTAAGTTTGCCGGAAAGATGTGATCCTCTTCCGCTGCTGAAATCAATATTGGTTTCTTGCTATCAATTGCTTCAATTTCCTCGATGCTGACGAAAGATGGATGTGCAATGGCTGCAGCATTGGCAAGACCCCCGTCGCCACTAATGTGTTGGACGGCAAACTTTGCACCAAAACAGTAACCCACAACGCCAATAAACTTTGGGTCATATTCAAGTTTTAACAACTTCATGAATCCATCAACAATTTTCTTGGTGACTTCAGGAGAATGTCTTTGAAACCAGGCATCACGATCAATTGGTTTGTCCGATGAGATAGCATCGCCGAATAAAATATCGGGAACAAAGACCATGTACCCAGCACTAGCAAATTTGTCGGCCGTTAATAAAACATTGTTGAATTTATTGCCATACACATCTGTCAAGATAACTATAACTTTTTCCTTGGGAGATGTAGAGCCTGCTGCATAAGTATCTAAACCGAAGATTTCTTCACGACGACCCTTGGGTGTTCCATCGTGACAAACTCCTTCAAAGCAACACTTGCCAGGTTGATTAGATGCCATTTGATTGAAATAATTTTGTTCTGCTGTAGTTAGACGTAGTGGAAAACTTTAAGTCTACTGAGTCTTGAGACCTTATCACCCTTTGAAGGTTTCTTGCAAATGAGCGTGGTTT\n",
      " \n",
      "What do you want to do to this gene?\n",
      "\"1\" delete it cleanly\n",
      "\"2\" replace it with an ORF/pre-made fragment\n",
      "\"3\" replace it with a standard cassette I will build for you\n",
      "\"4\" replace a specified region near your target gene\n",
      "Your answer: 1\n",
      " \n",
      "Hmm.. that's a bold choice. But OK, I'll design it for you.\n",
      "These are what you want stitched together, in order, I think:\n",
      " \n",
      "UpHom\n",
      "DownHom\n",
      " \n",
      "Now we're cooking with gas! I have some primers for you:\n",
      "Lup ID: UpHom\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('GTGAATTTGATCAATGCATTTTTCTTTTTG', SingleLetterAlphabet())\n",
      "Rup TAGGTAGGAGTTTGGAAAGAAAATTTGACTTACGCAACCTAATCTATTATTAACCTGCT\n",
      "Ldown TTAATAATAGATTAGGTTGCGTAAGTCAAATTTTCTTTCCAAACTCCTACCTATTCATTTC\n",
      "Rdown ID: <unknown id>\n",
      "Name: <unknown name>\n",
      "Description: <unknown description>\n",
      "Number of features: 0\n",
      "Seq('AAACCACGCTCATTTGCAAGAAACCTTCAA', SingleLetterAlphabet())\n",
      " \n",
      "\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python\n",
    "#title           : yCRISPRv1.7 \n",
    "#description     : defines several methods to generate oligos for CRISPR\n",
    "#author          : Leo d'Espaux <leodespaux@gmail.com>\n",
    "#date            : 9 June 2015\n",
    "#==============================================================================\n",
    "\n",
    "\n",
    "\n",
    "# import libraries we're using\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqUtils import MeltingTemp\n",
    "from Bio import SeqIO\n",
    "from intermine.webservice import Service\n",
    "\n",
    "\n",
    "# define global variables\n",
    "HomologyLength=1000 \n",
    "\n",
    "\n",
    "\n",
    "def askUser():\n",
    "    print(\"Hi baby, what are you in the mood for today? Using quotes enter\")\n",
    "    print(\"\\\"1\\\" to insert to an empty locus\") \n",
    "    print(\"\\\"2\\\" to edit an existing gene\")\n",
    "    print(\" \")\n",
    "\n",
    "    Action= input(\"Your answer: \")\n",
    "    if Action == \"1\":\n",
    "        print(\" \")\n",
    "        print(\"Do you want to construct a cassette?\")\n",
    "        print(\"\\\"1\\\" to construct using standard promoter and terminator names, and a custom sequence\")\n",
    "        print(\"\\\"2\\\" to construct using standard promoter and terminator names, and a custom sequence\")\n",
    "        print(\" \")\n",
    "\n",
    "        typeEdit=input(\"Your answer: \")\n",
    "        if typeEdit==\"1\":\n",
    "            print(\"Coming\")\n",
    "        inserttoEmpty(typeEdit)\n",
    "    elif Action == \"2\":\n",
    "        editExisting()\n",
    "    \n",
    "                # promt and ask to pick between these two options\n",
    "                # 1 == \"insert a gene into a new locus\"\n",
    "                # 2 == \"edit exiting locus\n",
    "                # and go to either of the next two functios\n",
    "                     \n",
    "                    \n",
    "                    \n",
    "def buildCassette():\n",
    "    print(\"Coming soon.\")        \n",
    "\n",
    "     \n",
    "    \n",
    "    \n",
    "def inserttoEmpty(Sequence):\n",
    "    print(\"Which of these characterized loci is going to get lucky: \")\n",
    "    print(\" \")\n",
    "    print(\"208a, the first we used, defined as 100% \")    \n",
    "    print(\"1021b, 105% \")\n",
    "    print(\" \")\n",
    "    print(\"You know the drill, all answers in quotes.\")\n",
    "    \n",
    "    filledLocus=input(\"Your answer: \")\n",
    "    \n",
    "    # find locus info to primers\n",
    "    # promt a table  \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "def editExisting():\n",
    "    print(\" \")\n",
    "    print(\"OK sweetie. Which locus do you want to edit, e.g., \\\"OAF1\\\": \")\n",
    "    print(\"Right now I only have chromosome 1 for testing.\")\n",
    "\n",
    "    \n",
    "    GeneName= input(\"Your answer: \")\n",
    "    \n",
    "    OrigGeneRecord=fetchGene(GeneName)\n",
    "    #note that this returns a seqrecord\n",
    "    \n",
    "    print(\" \")\n",
    "    print(\"I'm fetching upstream and downstream homology reghhions for you...\")\n",
    "    print(\" \")\n",
    "\n",
    "    \n",
    "    # let's load the appropriate chromosome file. The record of the gene we looked up\n",
    "    # contains in the \"features\" the systematic name, wherein the second letter\n",
    "    # corresponds to chromosome number, e.g., 1=A etc\n",
    "    if OrigGeneRecord.features[1]==\"A\":\n",
    "        origChromosomeRec=SeqIO.read(\"Scer01.fasta\", \"fasta\")\n",
    "    \n",
    "    # let's explicitely name the sequences from the seq record\n",
    "    OrigGeneSeq=OrigGeneRecord.seq\n",
    "    OrigChromosomeSeq=origChromosomeRec.seq\n",
    "    \n",
    "    # flip the sequence to orient with respect to the old gene\n",
    "    if OrigChromosomeSeq.find(OrigGeneSeq)==-1:\n",
    "        OrigChromosomeSeq=OrigChromosomeSeq.reverse_complement()\n",
    "\n",
    "        print (\"OK. I'm done. You might want to know that your gene is in the Crick strand.\")\n",
    "        print (\"Don't worry about who Crick is, though. I flipped your chromosome so you don't get confused.\")\n",
    "        \n",
    "    # Note that fetchHomology takes seq as inputs for gene and locus\n",
    "    UpHomSeq=fetchHomology(OrigGeneSeq,OrigChromosomeSeq,\"upstream\",HomologyLength)\n",
    "    DownHomSeq=fetchHomology(OrigGeneSeq,OrigChromosomeSeq,\"downstream\",HomologyLength)\n",
    "\n",
    "    \n",
    "    UpHomRec = SeqRecord(UpHomSeq, id=\"UpHom\")\n",
    "    DownHomRec = SeqRecord(DownHomSeq, id=\"DownHom\")\n",
    "    \n",
    "    print(\"OK I found them.\")\n",
    "\n",
    "    print(\" \")\n",
    "    print(\"UpstreamHomology reghhion is\")\n",
    "    print(UpHomRec.seq)\n",
    "    \n",
    "    print(\" \")\n",
    "    print(\"DownstreamHomology reghhion is\")\n",
    "    print(DownHomRec.seq)\n",
    "    \n",
    "    print(\" \")\n",
    "    print(\"What do you want to do to this gene?\")\n",
    "    print(\"\\\"1\\\" delete it cleanly\")\n",
    "    print(\"\\\"2\\\" replace it with an ORF/pre-made fragment\")\n",
    "    print(\"\\\"3\\\" replace it with a standard cassette I will build for you\")\n",
    "    print(\"\\\"4\\\" replace a specified region near your target gene\")\n",
    "    \n",
    "    Action=input(\"Your answer: \")\n",
    "                        \n",
    "    #note that in all the below, we want to have fragments be records\n",
    "    if Action==\"1\":\n",
    "        fragments=[UpHomRec,DownHomRec]\n",
    "    \n",
    "    if Action==\"2\":\n",
    "        print(\" \")\n",
    "        NewGeneName=input(\"What's the name of the gene you're inserting?\")\n",
    "        NewGeneSeq=Seq(input(\"What's the sequence of your new gene?\"))\n",
    "        InsertRec = SeqRecord(NewGeneSeq, id=NewGeneName)\n",
    "        fragments=[UpHomRec, InsertRec, DownHomRec]\n",
    "\n",
    "    if Action==\"3\":\n",
    "        print(\" \")\n",
    "        print(\"I got you baby. I don't want you to strain your eyes looking up stuff.\")\n",
    "        print(\"Actually I haven't yet gotten to that so no insert.\")\n",
    "\n",
    "        InsertRec=buildCassette()\n",
    "        fragments=[UpHomRec, DownHomRec] #we need to finish buildcassette to add InsertRec here\n",
    "\n",
    "    if Action==\"4\":\n",
    "        print(\"Actually I haven't yet gotten to that so no insert.\")\n",
    "        fragments=[UpHomRec, DownHomRec] #we need to finish buildcassette to add InsertRec here\n",
    "\n",
    "        \n",
    "        \n",
    "    # note that fragments are seqrecords\n",
    "\n",
    "    stitchTogether(fragments)\n",
    "    \n",
    "    print(\"\")\n",
    "    \n",
    "    \n",
    "\"\"\"\n",
    "Stitches together fragments and returns primers to do so. \n",
    "@param fragments SeqRecord List of [up insert down] or [up down] \n",
    "\"\"\"    \n",
    "def stitchTogether(fragments):\n",
    "    #this function takes seq records and prints primers\n",
    "    #at some point maybe a .gb file should be written\n",
    "    print(\" \")\n",
    "    print(\"Hmm.. that's a bold choice. But OK, I'll design it for you.\")\n",
    "    print(\"These are what you want stitched together, in order, I think:\")\n",
    "\n",
    "    print(\" \")\n",
    "    for i in fragments:\n",
    "        print(i.id)\n",
    "\n",
    "    print(\" \")\n",
    "    print(\"Now we're cooking with gas! I have some primers for you:\")\n",
    "    \n",
    "    up = fragments[0]\n",
    "    insert = fragments[1]\n",
    "    # If deletion is occuring, reassign down as insert\n",
    "    if len(fragments) == 2:\n",
    "        down = insert\n",
    "    else:\n",
    "        down = fragments[2]\n",
    "\n",
    "\n",
    "    Rtemp1 = up.reverse_complement()\n",
    "    Rtemp2 = insert.reverse_complement()\n",
    "\n",
    "    rPrim, rLength = getPrimer(Rtemp1)\n",
    "    lPrim, lLength = getPrimer(down)\n",
    "\n",
    "    Rup = getOverhang(Rtemp2, rLength) + rPrim\n",
    "    if len(fragments) == 2:\n",
    "        Ldown = getOverhang(up, lLength) + lPrim\n",
    "    else:\n",
    "        Ldown = getOverhang(insert, lLength) + lPrim\n",
    "    \n",
    " \n",
    "    Lup = up[:30]\n",
    "    Rdown = down[-30:].reverse_complement()\n",
    "\n",
    "    # Lup and Rdown are actually SeqRecords and must be converted to seq and then str \n",
    "    # to print nicely. \n",
    "\n",
    "    print(\"Lup\" + \" \" + str(Lup.seq))\n",
    "    print(\"Rup\" + \" \" + Rup)\n",
    "    print(\"Ldown\" + \" \" + Ldown)\n",
    "    print(\"Rdown\" + \" \" + str(Rdown.seq))\n",
    "    \n",
    "    print(\" \")\n",
    "    \n",
    "\"\"\"\n",
    "Helper method for stitchTogether\n",
    "Gets the primer from a given sequence based on melting temperature and max length. \n",
    "@param seq A sequence (or string) that represents the \n",
    "       sequence from which to derive the primer. The primer\n",
    "       is generated from left to right\n",
    "\"\"\"\n",
    "\n",
    "def getPrimer(seq):\n",
    "    \n",
    "    Tmax = 57\n",
    "    maxLen = 60\n",
    "    overhang = 0.8\n",
    "\n",
    "    mp = 0\n",
    "    length = 0\n",
    "    primer = Seq(\"\")\n",
    "\n",
    "    while mp <= Tmax and length <= maxLen / (1 + overhang):\n",
    "        primer = primer + seq[length]\n",
    "        mp = MeltingTemp.Tm_NN(primer)\n",
    "        length += 1\n",
    "\n",
    "    lastThree = primer[-3:]\n",
    "    while 'A' not in lastThree and 'T' not in lastThree:\n",
    "        primer = primer[:-1]\n",
    "        lastThree = primer[-3:]\n",
    "\n",
    "    return primer, length\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Helper method for stitchTogether\n",
    "Gets the overhang from a given sequence based on non-overhang length. \n",
    "@param seq Sequence from which to derive overhang. overhang is \n",
    "       generated from right to left\n",
    "@param pLength Length of the non-overhang part of the primer. \n",
    "\"\"\"\n",
    "def getOverhang(seq, pLength):\n",
    "\n",
    "    maxLen = 60\n",
    "    overhang = 0.8\n",
    "\n",
    "    length = 1\n",
    "    oh = Seq(\"\")\n",
    "    while length <= overhang * pLength:\n",
    "        oh = seq[-length] + oh\n",
    "        length += 1\n",
    "    return oh\n",
    "\n",
    "\n",
    "    #------------------------------ FETCH FUNCTIONS -------------------------------------\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "def fetchGene(GeneName):\n",
    "    \n",
    "    service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
    "    template = service.get_template('Gene_GenomicDNA')\n",
    "\n",
    "    rows = template.rows(\n",
    "        E = {\"op\": \"LOOKUP\", \"value\": GeneName, \"extra_value\": \"S. cerevisiae\"}\n",
    "    )\n",
    "    \n",
    "    # this service seems to return multiple similar genes but we want the first one only, so count\n",
    "    # and it returns information about the gene you want\n",
    "    count=0\n",
    "    for row in rows:\n",
    "        \n",
    "        count=count+1\n",
    "        if count==1:\n",
    "            descr= row[\"description\"]\n",
    "            GeneSeq=Seq(row[\"sequence.residues\"])\n",
    "            GeneSysName=row[\"secondaryIdentifier\"]\n",
    "            print(\" \")\n",
    "            print(\"I think you want...... \"+row[\"secondaryIdentifier\"])\n",
    "            print(row[\"description\"])\n",
    "            print(\" \")\n",
    "            print(row[\"sequence.residues\"])\n",
    "            print(\" \")\n",
    "            print(\"Good choice! I have a feeling you're going to get lucky with this one.\")\n",
    "            print(\" \")\n",
    "            print(\"Give me a second to put some of my ducks in a circle...\")\n",
    "       \n",
    "\n",
    "            \n",
    "    #let's create a record for the oldGene\n",
    "    GeneRecord = SeqRecord(GeneSeq, id=GeneSysName)\n",
    "    \n",
    "    #now let's add some more information to make it useful\n",
    "    GeneRecord.name=GeneName\n",
    "    GeneRecord.features=GeneSysName\n",
    "\n",
    "    return GeneRecord\n",
    "    \n",
    "  \n",
    "    \n",
    "\n",
    "    \n",
    "def fetchHomology(OrigGeneSeq,OrigChromosomeSeq,direction,HomologyLength):\n",
    "    # This function takes seq inputs for gene and locus, and also returns a seq\n",
    "\n",
    "    StartIndex=OrigChromosomeSeq.find(OrigGeneSeq)\n",
    "    EndIndex=StartIndex+len(OrigGeneSeq)\n",
    "    \n",
    "    if direction==\"upstream\":\n",
    "        HomologySeq=OrigChromosomeSeq[StartIndex-HomologyLength:StartIndex]\n",
    "    if direction==\"downstream\":\n",
    "        HomologySeq=OrigChromosomeSeq[EndIndex:EndIndex+HomologyLength]\n",
    "\n",
    "    return HomologySeq\n",
    "    # note that these are returned as an array of .seq \n",
    "\n",
    "    \n",
    "    \n",
    "    #allRecords=[SeqIO.read(\"Yali0A_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0B_contig.fasta\", \"fasta\"),\n",
    "      #  SeqIO.read(\"Yali0C_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0D_contig.fasta\", \"fasta\"),\n",
    "    #    SeqIO.read(\"Yali0E_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0F_contig.fasta\", \"fasta\")]\n",
    "        \n",
    "       # for records in allRecords:\n",
    "            # record.find(input)\n",
    "            # return chromosome, id, direction. if direction is different,\n",
    "            # return relative to the original ID, i.e., flip the chromosome view\n",
    "            \n",
    "     \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    #------------------------------------ CONSTRUCTING STUFF --------------------------------------\n",
    "\n",
    "    \n",
    "\n",
    "        \n",
    "def makeCassette(construction):\n",
    "    print(\"I can help you with that. Don't worry, I won't take your job. You are smart and beautiful.\")\n",
    "    print(\"Do you want to type the name of the promoter? e.g. \\\"TEF1p\\\"? If not, type \\\"custom\\\"\")\n",
    "    return SeqRecord(Seq(\"ATCG\",id=\"test\"))\n",
    "\n",
    "  \n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "askUser()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
