{
 "metadata": {
  "name": "",
  "signature": "sha256:42633067449b42ec295b34d2c7dee2b4cb716c810c3e721a640bda4c1dae34e4"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#!/usr/bin/env python\n",
      "#title           : yCRISPRv1.4.py\n",
      "#description     : defines several methods to generate oligos for CRISPR\n",
      "#author          : Leo d'Espaux <leodespaux@gmail.com>, help from Xingkai Li\n",
      "#date            : 9 June 2015\n",
      "#version         : 3.0\n",
      "#notes           : includes options for \n",
      "#notes           : delGene replGene cutOligos\n",
      "#notes           : at some point there will be one interface\n",
      "#==============================================================================\n",
      "\n",
      "\n",
      "\n",
      "# import libraries we're using\n",
      "from Bio.Seq import Seq\n",
      "from intermine.webservice import Service\n",
      "\n",
      "\n",
      "def askUser():\n",
      "    print(\"Hi baby, what are you in the mood for today? Using quotes enter\")\n",
      "    print(\"\\\"1\\\" to insert to an empty locus, \\\"2\\\" to edit an existing gene: \")\n",
      "    print(\" \")\n",
      "\n",
      "    Action=input()\n",
      "    if Action == \"1\":\n",
      "        print(\" \")\n",
      "        print(\"Do you want to construct a cassette?\")\n",
      "        print(\"\\\"1\\\" to construct using standard promoter and terminator names, and a custom sequence\")\n",
      "        print(\"\\\"2\\\" to construct using standard promoter and terminator names, and a custom sequence\")\n",
      "        print(\" \")\n",
      "\n",
      "        typeEdit=input()\n",
      "        if typeEdit==\"1\":\n",
      "            print(\"Coming\")\n",
      "        inserttoEmpty(typeEdit)\n",
      "    elif Action == \"2\":\n",
      "        editExisting()\n",
      "    \n",
      "                # promt and ask to pick between these two options\n",
      "                # 1 == \"insert a gene into a new locus\"\n",
      "                # 2 == \"edit exiting locus\n",
      "                # and go to either of the next two functios\n",
      "                 \n",
      "def makeCassette():\n",
      "    print(\"Coming soon.\")\n",
      "\n",
      "def inserttoEmpty(Sequence):\n",
      "    print(\"Which of these characterized loci is going to get lucky: \")\n",
      "    print(\" \")\n",
      "    print(\"208a, the first we used, defined as 100% \")    \n",
      "    print(\"1021b, 105% \")\n",
      "    print(\" \")\n",
      "    print(\"You know the drill, all answers in quotes.\")\n",
      "    \n",
      "    filledLocus=input()\n",
      "    \n",
      "    # find locus info to primers\n",
      "    # promt a table\n",
      "        \n",
      "    \n",
      "    \n",
      "def editExisting():\n",
      "    print(\" \")\n",
      "    print(\"OK sweetie which locus do you want to edit, quotes and caps as in \\\"GPD1\\\": \")\n",
      "    \n",
      "    GeneName=input()\n",
      "    \n",
      "    service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
      "    template = service.get_template('Gene_GenomicDNA')\n",
      "\n",
      "    rows = template.rows(\n",
      "        E = {\"op\": \"LOOKUP\", \"value\": GeneName, \"extra_value\": \"S. cerevisiae\"}\n",
      "    )\n",
      "    \n",
      "    # this service seems to return multiple similar genes but we want the first one only, so count\n",
      "    count=0\n",
      "    for row in rows:\n",
      "        \n",
      "        count=count+1\n",
      "        if count==1:\n",
      "            descr= row[\"description\"]\n",
      "            GeneSeq=Seq(row[\"sequence.residues\"])\n",
      "            print(\" \")\n",
      "            print(\"I think you want...... \"+row[\"secondaryIdentifier\"])\n",
      "            print(row[\"description\"])\n",
      "            print(\" \")\n",
      "            print(row[\"sequence.residues\"])\n",
      "            print(\" \")\n",
      "            print(\"I have a feeling you're going to get lucky with this one... do you want to: \")\n",
      "            print(\" \")\n",
      "            print(\"\\\"1\\\" just delete the gene, ASAP\")\n",
      "            print(\"\\\"2\\\" replace the old ORF with another ORF, or custom sequence\")\n",
      "            print(\"\\\"3\\\" it's complicated, replace a custom nearby sequence with another custom sequence\")\n",
      "            \n",
      "            __init__(GeneName, GeneSeq, id=GeneName, name='<unknown name>', \n",
      "                     description='<unknown description>', dbxrefs=None, \n",
      "                     features=None, annotations=None, letter_annotations=None)\n",
      "            \n",
      "            oldGene=Seq(rpw[\"sequence.residues\"])\n",
      "            oldGene.id\n",
      "            oldGeneName=\n",
      "        upstream=findDNA(systematic name, gene, \"up\")\n",
      "        downstream=findDNA()\n",
      "        \n",
      "        typeofEdit=input()\n",
      "        if typeofEdit==\"1\":\n",
      "            delGene(oldGene)\n",
      "            # in deleteGene, update to look up the chromosome\n",
      "    oldGeneIndex=findDNA(oldGene)\n",
      "    \n",
      "\n",
      "def findDNA(input):\n",
      "    # this program loads on-machine chromosomal data from jun 2015 and finds a sequence within it\n",
      "    # it returns\n",
      "    # chromosomeNum, index, plus or minus\n",
      "    # loading Chromosomes\n",
      "    \n",
      "    allRecords=[SeqIO.read(\"Yali0A_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0B_contig.fasta\", \"fasta\"),\n",
      "        SeqIO.read(\"Yali0C_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0D_contig.fasta\", \"fasta\"),\n",
      "        SeqIO.read(\"Yali0E_contig.fasta\", \"fasta\"), SeqIO.read(\"Yali0F_contig.fasta\", \"fasta\")]\n",
      "        \n",
      "       # for records in allRecords:\n",
      "            # record.find(input)\n",
      "            # return chromosome, id, direction. if direction is different,\n",
      "            # return relative to the original ID, i.e., flip the chromosome view\n",
      "            \n",
      "     \n",
      "    \n",
      "    \n",
      "\n",
      "\n",
      "\n",
      "def fetchGene(GeneName):\n",
      "# Here we fetch genomic DNA using a name and return the sequence.\n",
      "# If this works, you should see some text describing the gene as well.\n",
      "    \n",
      "    service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
      "    template = service.get_template('Gene_GenomicDNA')\n",
      "\n",
      "    rows = template.rows(\n",
      "        E = {\"op\": \"LOOKUP\", \"value\": GeneName, \"extra_value\": \"S. cerevisiae\"}\n",
      "    )\n",
      "    \n",
      "    # this service seems to return multiple similar genes but we want the first one only, so count\n",
      "    count=0\n",
      "    for row in rows:\n",
      "        \n",
      "        count=count+1\n",
      "        if count==1:\n",
      "            descr= row[\"description\"]\n",
      "            old=Seq(row[\"sequence.residues\"])\n",
      "            print(\" \")\n",
      "            print(\"I think you want...... \"+row[\"secondaryIdentifier\"])\n",
      "            print(row[\"description\"])\n",
      "            print(\" \")\n",
      "\n",
      "            return Seq(row[\"sequence.residues\"])\n",
      "\n",
      "    \n",
      "    \n",
      "def fetchHomology(GeneName, Direction):\n",
      "    from intermine.webservice import Service\n",
      "    service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
      "\n",
      "    # For a given gene(s), get a chosen length of upstream and/or downstream\n",
      "    # sequence along with the gene sequence. Choose to retrieve 0.5 kb, 1.0 kb, or\n",
      "    # 2.0 kb of upstream or downstream flanking regions, or both. To retrieve only\n",
      "    # flanking regions without the coding sequence, select \u201cfalse\u201d for the \u201cGene\n",
      "    # Flanking Region > Include Gene\u201d option (for this option, you must select only\n",
      "    # \u201cupstream\u201d or \u201cdownstream\u201d flanking regions, not \u201cboth\u201d).\n",
      "\n",
      "    from intermine.webservice import Service\n",
      "    service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
      "\n",
      "    # Retrieve the chromosomal location for a specified <a href =\n",
      "    # \"http://www.yeastgenome.org/yeastmine-help-page#gene\">gene</a>.\n",
      "\n",
      "    template = service.get_template('Gene_ChromosomeLocation')\n",
      "\n",
      "    # You can edit the constraint values below\n",
      "    # A    Gene\n",
      "\n",
      "    rows = template.rows(\n",
      "        A = {\"op\": \"LOOKUP\", \"value\": \"act1\"}\n",
      "    )\n",
      "    for row in rows:\n",
      "        print row[\"primaryIdentifier\"], row[\"secondaryIdentifier\"], row[\"symbol\"], row[\"name\"], \\\n",
      "            row[\"organism.shortName\"], row[\"chromosome.primaryIdentifier\"], \\\n",
      "            row[\"chromosomeLocation.start\"], row[\"chromosomeLocation.end\"], \\\n",
      "            row[\"chromosomeLocation.strand\"]\n",
      "        \n",
      "def stitchFragments(fragments):\n",
      "    print(\"Working on it\" )\n",
      "    #fragments is an array\n",
      "        \n",
      "        \n",
      "        \n",
      "def cutOnly():\n",
      "    \n",
      "# This function asks user for a target cut site, and generates oligos to clone that cut site into \n",
      "# a CRISPR plasmid, and to screen for positive clones containing that cut site. It may also generate\n",
      "# universal homology regions to integrate or otherwise alter that target region. \n",
      "#\n",
      "# The current cut site architecture is:  >>>>> YtRNAp-HDV ribozyme- >20nt< -gRNA <<<<<\n",
      "\n",
      "    \n",
      "    GeneName=input(\"Name, using quotes: \")\n",
      "    cutsite=input(\"20-mer cut sequence, using quotes: \")\n",
      "    DNA=input(\"Locus sequence +/- a few kb, using quotes: \")\n",
      "\n",
      "    if DNA.find(cutsite)==-1:                   # If cutiste sequence found in ANTISENSE\n",
      "        DNA=Seq(DNA).reverse_complement()       # then reverse DNA, and turn it into a string\n",
      "    \n",
      "    index=DNA.find(cutsite)+16                  # index gives the start position of the string, e.g., 0. \n",
      "                                                # we add 16 since index+0=start of 20-mer, so index+16=cut site, \n",
      "                                                # 3 nt before last of 20mer\n",
      "        \n",
      "    Lup=DNA[index-520:index-490]                # This primer binds 500bp upstream of cut site\n",
      "\n",
      "    \n",
      "    cutSequence=Seq(\"cgggtggcgaatgggacttt\")+cutsite+Seq(\"gttttagagctagaaatagc\")\n",
      "    seqprimer=Seq(\"gacttt\")+cutsite\n",
      "    \n",
      "\n",
      "    print(\"cut\" + GeneName + \"  \" + cutSequence)\n",
      "    print(\"Lcolony\" + GeneName + \"  \" + seqprimer)\n",
      "    print(\"Lup\" + GeneName + \"  \" + Lup)\n",
      "  \n",
      "    \n",
      "    \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "def delGene(oldGene):\n",
      "\n",
      "# This function asks user for a chromosomal locus, a region to be deleted, a suitable CRIPSR cutsite \n",
      "# and outputs oligos for cloning of a pL308 Cas9-gRNA vector, and ones for generating a donor DNA\n",
      "# to delete the unwanted chromosomal region. Primers Lup+Rdown produce a 1kb band if deletion was\n",
      "# successful. \n",
      "# part of yCRISPRv3 by leodespaux@gmail.com\n",
      "\n",
      "    #\n",
      "    #GeneName \n",
      "\n",
      "    deletion=oldGene\n",
      "\n",
      "    print(\" \")\n",
      "    print(\"Ok I got you, do you want me to pick a cut site for you? \")\n",
      "    print(\"\\\"1\\\" or enter your own cut site:\")\n",
      "    print(\" \")\n",
      "    cutChoice=input()\n",
      "    if cutChoice==\"1\":\n",
      "        cutsite=cutFinder(oldGene)\n",
      "    else:\n",
      "        cutsite=Seq(cutChoice)\n",
      "\n",
      "\n",
      "\n",
      "    \n",
      "    if deletion.find(cutsite)==-1:\n",
      "        if deletion.reverse_complement().find(cutsite)==-1:\n",
      "            print (\"WARNING: Guide 20-mer sequence not found in deletion region.\")\n",
      "        \n",
      "    locus=Seq(locus)\n",
      "    \n",
      "    index=locus.find(deletion)                  \n",
      "    # index gives the start position within locus of the region to be deleted\n",
      "    # note that it is\n",
      "    \n",
      "    newlocus=locus[0:index+1]+locus[index+len(deletion)-1:len(locus)]\n",
      "\n",
      "    # note that since index starts at 0, a value of n points to, in the newlocus,\n",
      "    # the first nt after the deletion. So we define the newlocus as above. Note too\n",
      "    # that a string of len=40 ends at an index of 39--so we pick up at index+len-1. \n",
      "    \n",
      "    Lup=newlocus[index-500:index-470]\n",
      "    Rup=newlocus[index-31:index+29].reverse_complement()\n",
      "    \n",
      "    Ldown=newlocus[index-31:index+29]\n",
      "    Rdown=newlocus[index+469:index+499].reverse_complement()\n",
      "    \n",
      "    cutSequence=Seq(\"cgggtggcgaatgggacttt\")+cutsite+Seq(\"gttttagagctagaaatagc\")\n",
      "    seqprimer=Seq(\"gacttt\")+cutsite\n",
      "    \n",
      "    print(\"cut\" + GeneName + \"  \" + cutSequence)\n",
      "    print(\"seq\" + GeneName + \"  \" + seqprimer)\n",
      "    print(\"Lup\" + GeneName + \"del\" + \" \" + Lup)\n",
      "    print(\"Rup\" + GeneName + \"del\" + \" \" + Rup)\n",
      "    print(\"Ldown\" + GeneName + \"del\" + \" \" + Ldown)\n",
      "    print(\"Rdown\" + GeneName + \"del\" + \" \" + Rdown)\n",
      "\n",
      "    \n",
      "    \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "def insertGene():\n",
      "    \n",
      "# This function asks user for a target cut site, a target locus, and a DNA sequence to be inserted.\n",
      "# The function outputs primers to allow one to PCR the insert with homology to the cut site. The final\n",
      "# integration happens 17nt from the cut site in both direction, such that a 34nt region is deleted.\n",
      "# The current cut site architecture is:  >>>>> YtRNAp-HDV ribozyme- >20nt< -gRNA <<<<<\n",
      "\n",
      "    locusName=input(\"Locus name, using quotes: \")\n",
      "    insertName=input(\"Name of gene to be inserted, using quotes: \")\n",
      "    \n",
      "    cutsite=input(\"20-mer cut sequence, using quotes: \").upper()\n",
      "    locus=input(\"Locus sequence +/- a few kb, using quotes: \")\n",
      "    insert=input(\"Enter insert DNA sequence, using quotes: \")\n",
      "\n",
      "\n",
      "    if locus.find(cutsite)==-1:                   # If cutsite sequence found in ANTISENSE\n",
      "        locus=Seq(locus).reverse_complement()     # then reverse DNA, and turn it into a string\n",
      "    \n",
      "    \n",
      "    index=locus.find(cutsite)                     # index gives the start position of cutsite within locus\n",
      "    \n",
      "        \n",
      "    Lprimer=locus[index-30:index]+insert[0:29]\n",
      "    Rprimer=Seq(insert[len(insert)-29:len(insert)]+locus[index+34:index+64]).reverse_complement()\n",
      "   \n",
      "    Lhom=locus[index-60:index]\n",
      "    Rhom=Seq(locus[index+34:index+94]).reverse_complement()\n",
      "    \n",
      " \n",
      "    print(\"L_\" + insertName + \"_:\" + locusName + \" \" + Lprimer)\n",
      "    print(\"R_\" + insertName + \"_:\" + locusName + \" \" + Rprimer)\n",
      "    print(\"Lext:\" + locusName + \" \" + Lhom)\n",
      "    print(\"Rext:\" + locusName + \" \" + Rhom)\n",
      "\n",
      "    \n",
      "    \n",
      "\n",
      "\n",
      "\n",
      "def replGene():\n",
      "    \n",
      "# This function asks user for a target cut site, a target locus, and a DNA sequence to be inserted.\n",
      "# The function outputs primers to allow one to PCR the insert with homology to the cut site. The final\n",
      "# integration happens 17nt from the cut site in both direction, such that a 34nt region is deleted.\n",
      "# The current cut site architecture is:  >>>>> YtRNAp-HDV ribozyme- >20nt< -gRNA <<<<<\n",
      "\n",
      "    locusName=input(\"Name of locus to be replaced, using quotes: \").lower()\n",
      "    insertName=input(\"Name of gene to be inserted, using quotes: \").upper()\n",
      "    cutsite=Seq(input(\"20-mer cut sequence, using quotes: \").upper())\n",
      "    locus=Seq(input(\"Locus sequence +/- a few kb, using quotes: \"))\n",
      "    deletion=Seq(input(\"DNA sequence to be replaced, using quotes: \"))\n",
      "    insert=Seq(input(\"Enter insert DNA sequence, using quotes: \"))\n",
      "\n",
      "\n",
      "   # if locus.find(cutsite)==-1:                   # If cutsite sequence found in ANTISENSE\n",
      "   #     locus=Seq(locus).reverse_complement()     # then reverse DNA, and turn it into a string\n",
      "    \n",
      "    index=locus.find(deletion)                     # index gives the start position of deletion\n",
      "        \n",
      "    # Let's define the locus resulting from the replacement  \n",
      "    newlocus=locus[0:index]+insert+locus[index+len(deletion):len(locus)]\n",
      "   \n",
      "    # primers to PCR insert\n",
      "    Lins= locus[index-30:index]+insert[0:30]\n",
      "    Rins= (insert[len(insert)-30:len(insert)]+locus[index+len(deletion):index+len(deletion)+30]).reverse_complement() \n",
      "     \n",
      "    # primers to PCR upstream homology region, ~500-bp\n",
      "    Lup=newlocus[index-500:index-470]\n",
      "    Rup=(locus[index-30:index]+insert[0:31]).reverse_complement() \n",
      "    \n",
      "    # And primers to PCR downstream homology region\n",
      "    Ldown=insert[len(insert)-30:len(insert)]+locus[index+len(deletion):index+len(deletion)+30]\n",
      "    Rdown=(locus[index+len(deletion)+470:index+len(deletion)+500]).reverse_complement()\n",
      "    \n",
      "    cutSequence=Seq(\"cgggtggcgaatgggacttt\")+cutsite+Seq(\"gttttagagctagaaatagc\")\n",
      "    seqprimer=Seq(\"gacttt\")+cutsite\n",
      "    \n",
      "    print(\"cut\" + locusName + \"  \" + cutSequence)\n",
      "    print(\"seq\" + locusName + \"  \" + seqprimer)\n",
      "    print(\"Lins_\" + locusName + \"::\"+insertName +\" \"+ Lins)\n",
      "    print(\"Rins_\" + locusName + \"::\"+insertName +\" \"+ Rins)\n",
      "    print(\"Lup_\" + locusName + \" \" + Lup)\n",
      "    print(\"Rup_\" + locusName + \"::\" + insertName +\" \" + Rup)\n",
      "    print(\"Ldown_\" + locusName + \"::\" + insertName +\" \" + Ldown)\n",
      "    print(\"Rdown_\" + locusName + \" \" + Rdown)\n",
      "    \n",
      "    \n",
      "    \n",
      "# to run, unhash your preferred program (and only that one) and ^Enter or play. E.g. remove hash below to run delGene()\n",
      "\n",
      "#cutOnly()\n",
      "#replGene()\n",
      "#insertGene()\n",
      "#delGene()\n",
      "\n",
      "\n",
      "def DoTwo():\n",
      "    a=input(\"Yo sup B :\" )\n",
      "    b=input(\"say what :\" )\n",
      "    return [a,b]\n",
      "\n",
      "  \n",
      "    \n",
      "askUser()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Hi baby, what are you in the mood for today? Using quotes enter\n",
        "\"1\" to insert to an empty locus, \"2\" to edit an existing gene: \n",
        " \n"
       ]
      },
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\"2\"\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " \n",
        "OK sweetie which locus do you want to edit, quotes and caps as in \"GPD1\": \n"
       ]
      },
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\"TDH1\"\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " \n",
        "So we are choosing...... YJL052W\n",
        "Glyceraldehyde-3-phosphate dehydrogenase (GAPDH), isozyme 1; involved in glycolysis and gluconeogenesis; tetramer that catalyzes the reaction of glyceraldehyde-3-phosphate to 1,3 bis-phosphoglycerate; detected in the cytoplasm and cell wall; protein abundance increases in response to DNA replication stress; GAPDH-derived antimicrobial peptides secreted by S. cerevisiae are active against a wide variety of wine-related yeasts and bateria\n",
        " \n",
        "ATGATCAGAATTGCTATTAACGGTTTCGGTAGAATCGGTAGATTGGTCTTGAGATTGGCTTTGCAAAGAAAAGACATTGAGGTTGTTGCTGTCAACGATCCATTTATCTCTAACGATTATGCTGCTTACATGGTCAAGTACGATTCTACTCATGGTAGATACAAGGGTACTGTTTCCCATGACGACAAGCACATCATCATTGATGGTGTCAAGATCGCTACCTACCAAGAAAGAGACCCAGCTAACTTGCCATGGGGTTCTCTAAAGATCGATGTCGCTGTTGACTCCACTGGTGTTTTCAAGGAATTGGACACCGCTCAAAAGCACATTGACGCTGGTGCCAAGAAGGTTGTCATCACTGCTCCATCTTCTTCTGCTCCAATGTTTGTTGTTGGTGTTAACCACACTAAATACACTCCAGACAAGAAGATTGTCTCCAACGCTTCTTGTACCACCAACTGTTTGGCTCCATTGGCCAAGGTTATCAACGATGCTTTCGGTATTGAAGAAGGTTTGATGACCACTGTTCACTCCATGACCGCCACTCAAAAGACTGTTGATGGTCCATCCCACAAGGACTGGAGAGGTGGTAGAACCGCTTCCGGTAACATTATCCCATCCTCTACCGGTGCTGCTAAGGCTGTCGGTAAGGTCTTGCCAGAATTGCAAGGTAAGTTGACCGGTATGGCTTTCAGAGTCCCAACCGTCGATGTTTCCGTTGTTGACTTGACTGTCAAGTTGGAAAAGGAAGCTACTTACGACCAAATCAAGAAGGCTGTTAAGGCTGCCGCTGAAGGTCCAATGAAGGGTGTTTTGGGTTACACCGAAGATGCCGTTGTCTCCTCTGATTTCTTGGGTGACACTCACGCTTCCATCTTCGATGCCTCCGCTGGTATCCAATTGTCTCCAAAGTTCGTCAAGTTGATTTCCTGGTACGATAACGAATACGGTTACTCCGCCAGAGTTGTTGACTTGATCGAATATGTTGCCAAGGCTTAA\n",
        " \n",
        "OK, I think you're going to get lucky with this one... do you want to: \n",
        " \n",
        "\"1\" just delete the gene, ASAP\n",
        "\"2\" replace the old ORF with another ORF, or custom sequence\n",
        "\"3\" it's complicated, replace a custom nearby sequence with another custom sequence\n"
       ]
      },
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\"1\"\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " \n",
        "Ok I got you, do you want me to pick a cut site for you? \n",
        "\"1\" or enter your own cut site:\n",
        " \n"
       ]
      },
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\"1\"\n"
       ]
      },
      {
       "ename": "NameError",
       "evalue": "global name 'cutFinder' is not defined",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
        "\u001b[0;32m<ipython-input-1-5a2288041017>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    387\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    388\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 389\u001b[0;31m \u001b[0maskUser\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
        "\u001b[0;32m<ipython-input-1-5a2288041017>\u001b[0m in \u001b[0;36maskUser\u001b[0;34m()\u001b[0m\n\u001b[1;32m     35\u001b[0m         \u001b[0minserttoEmpty\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtypeEdit\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     36\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mAction\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"2\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 37\u001b[0;31m         \u001b[0meditExisting\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     38\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m                 \u001b[0;31m# promt and ask to pick between these two options\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;32m<ipython-input-1-5a2288041017>\u001b[0m in \u001b[0;36meditExisting\u001b[0;34m()\u001b[0m\n\u001b[1;32m     95\u001b[0m         \u001b[0mtypeofEdit\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     96\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtypeofEdit\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"1\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 97\u001b[0;31m             \u001b[0mdelGene\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moldGene\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     98\u001b[0m             \u001b[0;31m# in deleteGene, update to look up the chromosome\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m     \u001b[0moldGeneIndex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfindDNA\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moldGene\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;32m<ipython-input-1-5a2288041017>\u001b[0m in \u001b[0;36mdelGene\u001b[0;34m(oldGene)\u001b[0m\n\u001b[1;32m    239\u001b[0m     \u001b[0mcutChoice\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    240\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mcutChoice\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"1\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 241\u001b[0;31m         \u001b[0mcutsite\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcutFinder\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moldGene\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    242\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    243\u001b[0m         \u001b[0mcutsite\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mSeq\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcutChoice\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;31mNameError\u001b[0m: global name 'cutFinder' is not defined"
       ]
      }
     ],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "input__init__(GeneName, GeneSeq, id=GeneName, name='<unknown name>', \n",
      "                     description='<unknown description>', dbxrefs=None, \n",
      "                     features=None, annotations=None, letter_annotations=None)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}