{
 "metadata": {
  "name": "",
  "signature": "sha256:f5a691e5a6f78c5828318213f29707e62298a2f397dfcc54c2379ebb5d11dd23"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#!/usr/bin/env python\n",
      "#title           : yCRISPRv3.py\n",
      "#description     : defines several methods to generate oligos for CRISPR\n",
      "#author          : Leo d'Espaux <leodespaux@gmail.com>, help from Xingkai Li\n",
      "#date            : 20150423\n",
      "#version         : 3.0\n",
      "#notes           : includes options for \n",
      "#notes           : delGene replGene cutOligos\n",
      "#notes           : at some point there will be one interface\n",
      "#==============================================================================\n",
      "\n",
      "from Bio.Seq import Seq\n",
      "\n",
      "\n",
      "\"\"\"\n",
      "Created on Thu Sep 11 16:01:49 2014\n",
      "@author: Leo leodespaux@gmail.com\n",
      "\"\"\"\n",
      "\n",
      "\n",
      "def cutOnly():\n",
      "    \n",
      "# This function asks user for a target cut site, and generates oligos to clone that cut site into \n",
      "# a CRISPR plasmid, and to screen for positive clones containing that cut site. It may also generate\n",
      "# universal homology regions to integrate or otherwise alter that target region. \n",
      "#\n",
      "# The current cut site architecture is:  >>>>> YtRNAp-HDV ribozyme- >20nt< -gRNA <<<<<\n",
      "\n",
      "    from Bio.Seq import Seq\n",
      "    GeneName=input(\"Name, using quotes: \")\n",
      "    cutsite=input(\"20-mer cut sequence, using quotes: \")\n",
      "    DNA=input(\"Locus sequence +/- a few kb, using quotes: \")\n",
      "\n",
      "    if DNA.find(cutsite)==-1:                   # If cutiste sequence found in ANTISENSE\n",
      "        DNA=Seq(DNA).reverse_complement()       # then reverse DNA, and turn it into a string\n",
      "    \n",
      "    index=DNA.find(cutsite)+16                  # index gives the start position of the string, e.g., 0. \n",
      "                                                # we add 16 since index+0=start of 20-mer, so index+16=cut site, \n",
      "                                                # 3 nt before last of 20mer\n",
      "        \n",
      "    Lup=DNA[index-520:index-490]                # This primer binds 500bp upstream of cut site\n",
      "\n",
      "    \n",
      "    cutSequence=Seq(\"cgggtggcgaatgggacttt\")+cutsite+Seq(\"gttttagagctagaaatagc\")\n",
      "    seqprimer=Seq(\"gacttt\")+cutsite\n",
      "    \n",
      "\n",
      "    print(\"cut\" + GeneName + \"  \" + cutSequence)\n",
      "    print(\"Lcolony\" + GeneName + \"  \" + seqprimer)\n",
      "    print(\"Lup\" + GeneName + \"  \" + Lup)\n",
      "  \n",
      "    \n",
      "    \n",
      "def delGene():\n",
      "\n",
      "# This function asks user for a chromosomal locus, a region to be deleted, a suitable CRIPSR cutsite \n",
      "# and outputs oligos for cloning of a pL308 Cas9-gRNA vector, and ones for generating a donor DNA\n",
      "# to delete the unwanted chromosomal region. Primers Lup+Rdown produce a 1kb band if deletion was\n",
      "# successful. \n",
      "# part of yCRISPRv3 by leodespaux@gmail.com\n",
      "\n",
      "    GeneName=input(\"Name, using quotes: \")\n",
      "    cutsite=input(\"20-mer cut sequence, using quotes: \").upper()\n",
      "    locus=input(\"Locus sequence +1few kb, using quotes: \")\n",
      "    deletion=input(\"Sequence to be deleted, using quotes: \")\n",
      "\n",
      "    deletion=Seq(deletion)\n",
      "    \n",
      "    if deletion.find(cutsite)==-1:\n",
      "        if deletion.reverse_complement().find(cutsite)==-1:\n",
      "            print (\"WARNING: Guide 20-mer sequence not found in deletion region.\")\n",
      "        \n",
      "    locus=Seq(locus)\n",
      "    \n",
      "    index=locus.find(deletion)                  \n",
      "   \n",
      "    # index gives the start position within locus of the string deletion.\n",
      "    # now we delete the deletion region to redefine a newlocus:\n",
      "    \n",
      "    newlocus=locus[0:index-1]+locus[index+len(deletion)-1:len(locus)]\n",
      "\n",
      "    # note that since index starts at 0, a value of n points to, in the newlocus,\n",
      "    # the first nt after the deletion. So we define the newlocus as above. Note too\n",
      "    # that a string of len=40 ends at an index of 39--so we pick up at index+len-1. \n",
      "    \n",
      "    Lup=newlocus[index-500:index-470]\n",
      "    Rup=newlocus[index-31:index+29].reverse_complement()\n",
      "    \n",
      "    Ldown=newlocus[index-31:index+29]\n",
      "    Rdown=newlocus[index+469:index+499].reverse_complement()\n",
      "    \n",
      "    cutSequence=Seq(\"cgggtggcgaatgggacttt\")+cutsite+Seq(\"gttttagagctagaaatagc\")\n",
      "    seqprimer=Seq(\"gacttt\")+cutsite\n",
      "    \n",
      "    print(\"cut\" + GeneName + \"  \" + cutSequence)\n",
      "    print(\"seq\" + GeneName + \"  \" + seqprimer)\n",
      "    print(\"Lup\" + GeneName + \"del\" + \" \" + Lup)\n",
      "    print(\"Rup\" + GeneName + \"del\" + \" \" + Rup)\n",
      "    print(\"Ldown\" + GeneName + \"del\" + \" \" + Ldown)\n",
      "    print(\"Rdown\" + GeneName + \"del\" + \" \" + Rdown)\n",
      "\n",
      "\n",
      "    \n",
      "\n",
      "    \n",
      "def insertGene():\n",
      "    \n",
      "# This function asks user for a target cut site, a target locus, and a DNA sequence to be inserted.\n",
      "# The function outputs primers to allow one to PCR the insert with homology to the cut site. The final\n",
      "# integration happens 17nt from the cut site in both direction, such that a 34nt region is deleted.\n",
      "# The current cut site architecture is:  >>>>> YtRNAp-HDV ribozyme- >20nt< -gRNA <<<<<\n",
      "\n",
      "    from Bio.Seq import Seq\n",
      "    locusName=input(\"Locus name, using quotes: \")\n",
      "    insertName=input(\"Name of gene to be inserted, using quotes: \")\n",
      "    \n",
      "    cutsite=input(\"20-mer cut sequence, using quotes: \").upper()\n",
      "    locus=input(\"Locus sequence +/- a few kb, using quotes: \")\n",
      "    insert=input(\"Enter insert DNA sequence, using quotes: \")\n",
      "\n",
      "\n",
      "    if locus.find(cutsite)==-1:                   # If cutsite sequence found in ANTISENSE\n",
      "        locus=Seq(locus).reverse_complement()     # then reverse DNA, and turn it into a string\n",
      "    \n",
      "    \n",
      "    index=locus.find(cutsite)                     # index gives the start position of cutsite within locus\n",
      "    \n",
      "        \n",
      "    Lprimer=locus[index-30:index]+insert[0:29]\n",
      "    Rprimer=Seq(insert[len(insert)-29:len(insert)]+locus[index+34:index+64]).reverse_complement()\n",
      "   \n",
      "    Lhom=locus[index-60:index]\n",
      "    Rhom=Seq(locus[index+34:index+94]).reverse_complement()\n",
      "    \n",
      " \n",
      "    print(\"L_\" + insertName + \"_:\" + locusName + \" \" + Lprimer)\n",
      "    print(\"R_\" + insertName + \"_:\" + locusName + \" \" + Rprimer)\n",
      "    print(\"Lext:\" + locusName + \" \" + Lhom)\n",
      "    print(\"Rext:\" + locusName + \" \" + Rhom)\n",
      "\n",
      "    \n",
      "    \n",
      "\n",
      "def replGene():\n",
      "    \n",
      "# This function asks user for a target cut site, a target locus, and a DNA sequence to be inserted.\n",
      "# The function outputs primers to allow one to PCR the insert with homology to the cut site. The final\n",
      "# integration happens 17nt from the cut site in both direction, such that a 34nt region is deleted.\n",
      "# The current cut site architecture is:  >>>>> YtRNAp-HDV ribozyme- >20nt< -gRNA <<<<<\n",
      "\n",
      "    from Bio.Seq import Seq\n",
      "    locusName=input(\"Name of locus to be replaced, using quotes: \").lower()\n",
      "    insertName=input(\"Name of gene to be inserted, using quotes: \").upper()\n",
      "    cutsite=Seq(input(\"20-mer cut sequence, using quotes: \").upper()\n",
      "    locus=Seq(input(\"Locus sequence +/- a few kb, using quotes: \"))\n",
      "    deletion=Seq(input(\"DNA sequence to be replaced, using quotes: \"))\n",
      "    insert=Seq(input(\"Enter insert DNA sequence, using quotes: \"))\n",
      "\n",
      "\n",
      "   # if locus.find(cutsite)==-1:                   # If cutsite sequence found in ANTISENSE\n",
      "   #     locus=Seq(locus).reverse_complement()     # then reverse DNA, and turn it into a string\n",
      "    \n",
      "    index=locus.find(deletion)                     # index gives the start position of deletion\n",
      "        \n",
      "    # Let's define the locus resulting from the replacement  \n",
      "    newlocus=locus[0:index]+insert+locus[index+len(deletion):len(locus)]\n",
      "   \n",
      "    # primers to PCR insert\n",
      "    Lins= locus[index-30:index]+insert[0:30]\n",
      "    Rins= (insert[len(insert)-30:len(insert)]+locus[index+len(deletion):index+len(deletion)+30]).reverse_complement() \n",
      "     \n",
      "    # primers to PCR upstream homology region, ~500-bp\n",
      "    Lup=newlocus[index-500:index-470]\n",
      "    Rup=(locus[index-30:index]+insert[0:30]).reverse_complement() \n",
      "    \n",
      "    # And primers to PCR downstream homology region\n",
      "    Ldown=insert[len(insert)-30:len(insert)]+locus[index+len(deletion):index+len(deletion)+30]\n",
      "    Rdown=(locus[index+len(deletion)+470:index+len(deletion)+500]).reverse_complement()\n",
      "    \n",
      "    cutSequence=Seq(\"cgggtggcgaatgggacttt\")+cutsite+Seq(\"gttttagagctagaaatagc\")\n",
      "    seqprimer=Seq(\"gacttt\")+cutsite\n",
      "    \n",
      "    print(\"cut\" + locusName + \"  \" + cutSequence)\n",
      "    print(\"seq\" + locusName + \"  \" + seqprimer)\n",
      "    print(\"Lins_\" + locusName + \"::\"+insertName +\" \"+ Lins)\n",
      "    print(\"Rins_\" + locusName + \"::\"+insertName +\" \"+ Rins)\n",
      "    print(\"Lup_\" + locusName + \" \" + Lup)\n",
      "    print(\"Rup_\" + locusName + \"::\" + insertName +\" \" + Rup)\n",
      "    print(\"Ldown_\" + locusName + \"::\" + insertName +\" \" + Ldown)\n",
      "    print(\"Rdown_\" + locusName + \" \" + Rdown)\n",
      "    \n",
      "    \n",
      "    \n",
      "# now to run, type the program and ^Enter or play. E.g. remove hash below to run delGene()\n",
      "\n",
      "replGene()\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Name of gene to be inserted, using quotes: \"fadM\"\n"
       ]
      },
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Enter insert DNA sequence, using quotes: \"ATGTCtcaaacccaaattaaagtaagaggttatcatctagatgtctaccagcacgtgaacaatgctagatatcttgaatttcttgaggaagctagatgggatggattagagaattccgactcttttcaatggatgactgctcataatattgcatttgtcgttgtcaacatcaatattaattacaggcgtcctgccgtcttgtccgatctattaactatcacatcacaattgcaacagttaaatggcaaaagtggcattctgagccaggtaataaccttagaaccagaaggtcaagtcgttgccgacgcattgatcacatttgtgtgtattgaccttaaaacccaaaaggccctggcattagaaggcgaattgagggaaaaattggaacaaatggttaaataa\"\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "cutPOT1  cgggtggcgaatgggactttAATGCATCCCAGGTCTCCGAgttttagagctagaaatagc\n",
        "seqPOT1  gactttAATGCATCCCAGGTCTCCGA\n",
        "Lins_POT1::FADM ACAGAGTGAATATAACACTACATAAAAGCAATGTCtcaaacccaaattaaagtaagaggt\n",
        "Rins_POT1::FADM GAGAATATTAACTATTATCAAGTATTAAAAttatttaaccatttgttccaatttttccct\n",
        "Lup_POT1 CAGCCCATCGTTCAAATTCAATTCTTCCAA\n",
        "Rup_POT1::FADM acctcttactttaatttgggtttgaGACATTGCTTTTATGTAGTGTTATATTCACTCTGT\n",
        "Ldown_POT1::FADM agggaaaaattggaacaaatggttaaataaTTTTAATACTTGATAATAGTTAATATTCTC\n",
        "Rdown_POT1 GGAGCTAATAAAAAGAATACCCTACACTAC\n"
       ]
      }
     ],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dd"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 0
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}